File: vcx/wrappers/python3/demo/alice.py
import asyncio
import json
from time import sleep

from vcx.api.connection import Connection
from vcx.api.credential import Credential
from vcx.api.disclosed_proof import DisclosedProof
from vcx.api.utils import vcx_agent_provision
from vcx.api.vcx_init import vcx_init_with_config
from vcx.state import State
from vc_auth_oidc.alice_vc_auth import handle_challenge


# logging.basicConfig(level=logging.DEBUG) uncomment to get logs

provisionConfig = {
    'agency_url': 'http://localhost:8080',
    'agency_did': 'VsKV7grR1BUE29mG2Fm2kX',
    'agency_verkey': 'Hezce2UWMZ3wUhVkh2LfKSs8nDzWwzs2Win7EzNN3YaR',
    'wallet_name': 'alice_wallet',
    'wallet_key': '123',
    'payment_method': 'null',
    'enterprise_seed': '000000000000000000000000Trustee1',
    'protocol_type': '4.0',
}


async def main():
    await init()
    connection_to_faber = None
    while True:
        answer = input(
            "Would you like to do? \n "
            "0 - establish connection \n "
            "1 - check for credential offer \n "
            "2 - check for proof request \n "
            "3 - pass vc_auth_oidc-challenge \n "
            "else finish \n") \
            .lower().strip()
        if answer == '0':
            connection_to_faber = await connect()
        elif answer == '1':
            print("Check agency for a credential offer")
            offers = await Credential.get_offers(connection_to_faber)
            credential = await Credential.create('credential', offers[0])
            await accept_offer(connection_to_faber, credential)
        elif answer == '2':
            print("Check agency for a proof request")
            requests = await DisclosedProof.get_requests(connection_to_faber)
            print("#23 Create a Disclosed proof object from proof request")
            proof = await DisclosedProof.create('proof', requests[0])
            await create_proof(connection_to_faber, proof)
        elif answer == '3':
            request = await handle_challenge()
            print("#23 Create a Disclosed proof object from proof request")
            proof = await DisclosedProof.create('proof', request)
            await create_proof(None, proof)
        else:
            break

    print("Finished")


async def init():
    print("#7 Provision an agent and wallet, get back configuration details")
    config = await vcx_agent_provision(json.dumps(provisionConfig))
    config = json.loads(config)
    # Set some additional configuration options specific to alice
    config['institution_name'] = 'alice'
    config['institution_logo_url'] = 'http://robohash.org/456'
    config['genesis_path'] = 'docker.txn'
    config['payment_method'] = 'null'

    config = json.dumps(config)

    print("#8 Initialize libvcx with new configuration")
    await vcx_init_with_config(config)


async def connect():
    print("#9 Input faber.py invitation details")
    details = input('invite details: ')

    print("#10 Convert to valid json and string and create a connection to faber")
    jdetails = json.loads(details)
    connection_to_faber = await Connection.create_with_details('faber', json.dumps(jdetails))
    await connection_to_faber.connect('{"use_public_did": true}')
    connection_state = await connection_to_faber.update_state()
    while connection_state != State.Accepted:
        sleep(2)
        await connection_to_faber.update_state()
        connection_state = await connection_to_faber.get_state()

    print("Connection is established")
    return connection_to_faber


async def accept_offer(connection_to_faber, credential):
    print("#15 After receiving credential offer, send credential request")
    await credential.send_request(connection_to_faber, 0)

    print("#16 Poll agency and accept credential offer from faber")
    credential_state = await credential.get_state()
    while credential_state != State.Accepted:
        sleep(2)
        await credential.update_state()
        credential_state = await credential.get_state()


async def create_proof(connection_to_faber, proof):
    print("#24 Query for credentials in the wallet that satisfy the proof request")
    credentials = await proof.get_creds()

    # Use the first available credentials to satisfy the proof request
    for attr in credentials['attrs']:
        credentials['attrs'][attr] = {
            'credential': credentials['attrs'][attr][0]
        }

    print("#25 Generate the proof")
    await proof.generate_proof(credentials, {})

    print("#26 Send the proof")
    await proof.send_proof(connection_to_faber)

    proof_state = await proof.get_state()
    while proof_state != State.Accepted:
        sleep(2)
        await proof.update_state()
        proof_state = await proof.get_state()

    print("proof is verified!!")


if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
    sleep(1)

File: vcx/wrappers/node/demo/alice.js
const { DisclosedProof } = require('../dist/src/api/disclosed-proof')
const { Connection } = require('../dist/src/api/connection')
const { Credential } = require('../dist/src/api/credential')
const { StateType } = require('../dist/src')
const readlineSync = require('readline-sync')
const sleepPromise = require('sleep-promise')
const demoCommon = require('./common')
const logger = require('./logger')
const url = require('url')
const isPortReachable = require('is-port-reachable')
const { runScript } = require('./script-comon')

const utime = Math.floor(new Date() / 1000)
const optionalWebhook = 'http://localhost:7209/notifications/alice'

const provisionConfig = {
  agency_url: 'http://localhost:8080',
  agency_did: 'VsKV7grR1BUE29mG2Fm2kX',
  agency_verkey: 'Hezce2UWMZ3wUhVkh2LfKSs8nDzWwzs2Win7EzNN3YaR',
  wallet_name: `node_vcx_demo_alice_wallet_${utime}`,
  wallet_key: '123',
  payment_method: 'null',
  enterprise_seed: '000000000000000000000000Trustee1'
}

const logLevel = 'error'

async function runAlice (options) {
  await demoCommon.initLibNullPay()

  logger.info('#0 Initialize rust API from NodeJS')
  await demoCommon.initRustApiAndLogger(logLevel)

  if (options.comm === 'aries') {
    provisionConfig.protocol_type = '4.0'
    logger.info('Running with Aries VCX Enabled! Make sure VCX agency is configured to use protocol_type 2.0')
  }

  if (options.postgresql) {
    await demoCommon.loadPostgresPlugin(provisionConfig)
    provisionConfig.wallet_type = 'postgres_storage'
    provisionConfig.storage_config = '{"url":"localhost:5432"}'
    provisionConfig.storage_credentials = '{"account":"postgres","password":"mysecretpassword","admin_account":"postgres","admin_password":"mysecretpassword"}'
    logger.info(`Running with PostreSQL wallet enabled! Config = ${provisionConfig.storage_config}`)
  } else {
    logger.info('Running with builtin wallet.')
  }
  if (await isPortReachable(url.parse(optionalWebhook).port, { host: url.parse(optionalWebhook).hostname })) { // eslint-disable-line
    provisionConfig.webhook_url = optionalWebhook
    logger.info(`Running with webhook notifications enabled! Webhook url = ${optionalWebhook}`)
  } else {
    logger.info('Webhook url will not be used')
  }

  logger.info('#8 Provision an agent and wallet, get back configuration details')
  const config = await demoCommon.provisionAgentInAgency(provisionConfig)

  logger.info('#9 Initialize libvcx with new configuration')
  await demoCommon.initVcxWithProvisionedAgentConfig(config)

  logger.info('Input faber.py invitation details')
  const details = readlineSync.question('Enter your invite details: ')
  const jdetails = JSON.parse(details)

  logger.info('#10 Convert to valid json and string and create a connection to faber')
  const connectionToFaber = await Connection.createWithInvite({ id: 'faber', invite: JSON.stringify(jdetails) })
  await connectionToFaber.connect({ data: '{"use_public_did": true}' })
  let connectionstate = await connectionToFaber.getState()
  while (connectionstate !== StateType.Accepted) {
    await sleepPromise(2000)
    await connectionToFaber.updateState()
    connectionstate = await connectionToFaber.getState()
  }

  logger.info('#11 Wait for faber.py to issue a credential offer')
  await sleepPromise(10000)
  const offers = await Credential.getOffers(connectionToFaber)
  logger.info(`Alice found ${offers.length} credential offers.`)
  logger.debug(JSON.stringify(offers))

  // Create a credential object from the credential offer
  const credential = await Credential.create({ sourceId: 'credential', offer: JSON.stringify(offers[0]) })

  logger.info('#15 After receiving credential offer, send credential request')
  await credential.sendRequest({ connection: connectionToFaber, payment: 0 })

  logger.info('#16 Poll agency and accept credential offer from faber')
  let credentialState = await credential.getState()
  while (credentialState !== StateType.Accepted) {
    await sleepPromise(2000)
    await credential.updateState()
    credentialState = await credential.getState()
  }

  logger.info('#22 Poll agency for a proof request')
  let requests = await DisclosedProof.getRequests(connectionToFaber)
  while (requests.length === 0) {
    await sleepPromise(2000)
    requests = await DisclosedProof.getRequests(connectionToFaber)
  }
  logger.info('#23 Create a Disclosed proof object from proof request')
  const proof = await DisclosedProof.create({ sourceId: 'proof', request: JSON.stringify(requests[0]) })

  logger.info('#24 Query for credentials in the wallet that satisfy the proof request')
  const credentials = await proof.getCredentials()

  // Use the first available credentials to satisfy the proof request
  for (let i = 0; i < Object.keys(credentials.attrs).length; i++) {
    const attr = Object.keys(credentials.attrs)[i]
    credentials.attrs[attr] = {
      credential: credentials.attrs[attr][0]
    }
    credentials.attrs[attr].tails_file = '/tmp/tails'
  }

  logger.info('#25 Generate the proof')
  await proof.generateProof({ selectedCreds: credentials, selfAttestedAttrs: {} })

  logger.info('#26 Send the proof to faber')
  await proof.sendProof(connectionToFaber)

  logger.info('#27 Wait for Faber to receive the proof')
  let proofState = await proof.getState()
  while (proofState !== StateType.Accepted && proofState !== StateType.None) {
    await sleepPromise(2000)
    await proof.updateState()
    proofState = await proof.getState()
  }
  logger.info('Faber received the proof')
  process.exit(0)
}

const optionDefinitions = [
  {
    name: 'help',
    alias: 'h',
    type: Boolean,
    description: 'Display this usage guide.'
  },
  {
    name: 'comm',
    type: String,
    description: 'Communication method. Possible values: aries, legacy. Default is aries.',
    defaultValue: 'aries'
  },
  {
    name: 'postgresql',
    type: Boolean,
    description: 'If specified, postresql wallet will be used.',
    defaultValue: false
  }
]

const usage = [
  {
    header: 'Options',
    optionList: optionDefinitions
  },
  {
    content: 'Project home: {underline https://github.com/Patrik-Stas/indy-wallet-watch}'
  }
]

function areOptionsValid (options) {
  const allowedCommMethods = ['aries', 'legacy']
  if (!(allowedCommMethods.includes(options.comm))) {
    console.error(`Unknown communication method ${options.comm}. Only ${JSON.stringify(allowedCommMethods)} are allowed.`)
    return false
  }
  return true
}

runScript(optionDefinitions, usage, areOptionsValid, runAlice)
